source/server/server.cc

```diff
@@ -287,6 +289,32 @@ void InstanceUtil::loadBootstrapConfig(envoy::config::bootstrap::v3::Bootstrap&
   MessageUtil::validate(bootstrap, validation_visitor);
 }
 
+static void loadSocketInterface(envoy::config::core::v3::TypedExtensionConfig sock_interface,
+                                ProtobufMessage::ValidationContext& validation_ctx) {
+  auto& factory =
+      Config::Utility::getAndCheckFactory<Network::SocketInterfaceFactory>(sock_interface);
+  auto config = Config::Utility::translateAnyToFactoryConfig(
+      sock_interface.typed_config(), validation_ctx.staticValidationVisitor(), factory);
+  Network::SocketInterfacesSingleton::get().emplace(
+      std::make_pair(sock_interface.name(), factory.createSocketInterface(*config)));
+}
+
+static void
+initializeSocketInterfaces(const envoy::config::core::v3::SocketInterfacesConfig& sock_interfaces,
+                           ProtobufMessage::ValidationContext& validation_ctx) {
+  Network::SocketInterfacesSingleton::initialize(new Network::SocketInterfacesMap());
+  for (const auto& sock_interface : sock_interfaces.custom_types()) {
+    loadSocketInterface(sock_interface, validation_ctx);
+  }
+  if (sock_interfaces.has_custom_default_type()) {
+    Network::SocketInterfaceSingleton::clear();
+    const auto& custom_sock = sock_interfaces.custom_default_type();
+    loadSocketInterface(custom_sock, validation_ctx);
+    auto sock = const_cast<Network::SocketInterface*>(Network::socketInterface(custom_sock.name()));
+    Network::SocketInterfaceSingleton::initialize(sock);
```

Comment:
In a perfect world we would move away from the scoped singleton pattern. Where do we need to access the interfaces(s)? Can we hang this off of `Api` perhaps?

Reply:
I know what you mean! :-) I wanted to do that instead of `SocketInterfaceSingleton` as well but then realized that we need to pass an `Api` instance down to `SocketImpl` through several layers of `Socket` derived classes (that's why I asked if we need that many types of sockets at one point). 

After merging `Socket` and `IoHandle`, if we stop relying on `Socket` constructors and instead directly allocate `Sockets` that we wrap in `TcpListenSocket`, `AcceptedSocketImpl` and the likes, then this might be even easier. That is, we should be able to do something like `api->socketInterface(name)->socket()` in those classes or those that allocate them. 

Would you like to try to do that with this patch or should we postpone it until we're done with the current set of changes/issues?

Reply:
Let's definitely tackle this in later changes. SGTM!